File: demos/Token_tps/Token.c
#include <stdio.h>
#include <stdlib.h>
// mode = {"create", "balance", "transfer"}
// parameter(name, "create", user, initial coin, path, execute amount): create an account
// parameter(name, "balance", user, path, execute amount): get balance
// parameter(name, "transfer", A, B, amount, path, execute amount) means A transfer ammount coins to B 
struct {
    char name[20];
    uint balance;
} Users;

void create(const int user, const int ini_coins, const char *path, int times) {
	int balance[3] = {0, 0, 0};
	FILE *fp = NULL;
	FILE *wfp = NULL;
	while (times-- > 0) {
		if ((fp = fopen(path, "a+")) == NULL) {
			printf("\nFirst open file");
		}
		for (int i = 0; i < 3; i++) {
			fscanf(fp, "%d\n", &balance[i]);
		}
		balance[user] = ini_coins;
		fclose(fp);
		if ((wfp = fopen(path, "w")) == NULL) {
			printf("\nCannot open write file");
		}
		for (int i = 0; i < 3; i++) {
			fprintf(wfp, "%d\n", balance[i]);
		}
		fclose(wfp);
		
	}
	// free(fp);
	// free(wfp);
}

void getBalance(const int user, const char *path, int times) {
	int balance[3] = {0, 0, 0};
	FILE *fp = NULL;
	while (times-- > 0) {
		if ((fp = fopen(path, "a+")) == NULL) {
			printf("\nFirst open file");
		}
		for (int i = 0; i < 3; i++) {
			fscanf(fp, "%d\n", &balance[i]);
		}
		printf("The balance of user %d is %d \n", user, balance[user]);
		fclose(fp);
		// free(fp);
	}
}

void transfer(const int user_a, const int user_b, const int value, const char *path, int times) {
	FILE *fp = NULL;
	FILE *wfp = NULL;
	while (times-- > 0) {
		int balance[3] = {0, 0, 0};
		if ((fp = fopen(path, "a+")) == NULL) {
			printf("\nFirst open file");
		}
		for (int i = 0; i < 3; i++) {
			fscanf(fp, "%d\n", &balance[i]);
		}
		fclose(fp);
		if ((wfp = fopen(path, "w")) == NULL) {
			printf("\nCannot open write file");
		}
		if (balance[user_a] < value) {
			printf("user has not enough coins!\n");
		}
		else {
			balance[user_a] -= value;
			balance[user_b] += value;
		}
	
		for (int i = 0; i < 3; i++) {
			fprintf(wfp, "%d\n", balance[i]);
		}
		fclose(wfp);
		// free(fp);
		// free(wfp);
	}
}

int main(int argc, char *argv[]) {
	char *mode;
    char *file_path;
    file_path = argv[1];

    char *user;
    user = argv[2];
    int ini_coins, times;
    // sscanf(argv[2], "%s", &user);
    sscanf(argv[3], "%d", &ini_coins);
    sscanf(argv[4], "%d", &times);
		// free(file_path);
    FILE *wfp = NULL;
    if ((wfp = fopen(file_path, "w")) == NULL) {
        printf("\nCannot open write file");
        return -1;
    }
    // printf("user is: %s", *user);
    for (int i = 0; i < times; i++) {
        fprintf(wfp, "%-20.20s %d\n", user, ini_coins);
        // fprintf(wfp, "%s\n", user);
    }
    fclose(wfp);
    
    return 0;
}

File: chainspace/for_testing/contracts/token.py
from chainspacecontract import ChainspaceContract

contract = ChainspaceContract('token')

@contract.method('init')
def init():
    return {
        'outputs': ('o',)
    }

@contract.method('create')
def create(inputs, reference_inputs, parameters):
    count = int(parameters[0])
    return {
        'outputs': ('o',)*count
    }

balance = [0, 0, 0]
@contract.method('create_account')
def create_account(inputs, reference_inputs, parameters):
    user = int(parameters[0])
    ini_coins = int(parameters[1])
    filepath = parameters[2]

    try:
        fp = open(filepath, "a+")
        for index in range(3):
            line = fp.next()
            balance[index] = int(line.strip())
    except StopIteration:
        fp.close()
    else:
        balance[user] = ini_coins
        fp.close()
    try:
        wfp = open(filepath, "w")
    except IOError:
        raise IOError("Cannot open write file")
    else:
        for i in range(3):
            wfp.write("{0}\n".format(balance[i]))
        wfp.close()

    return {}

@contract.method('getBalance')
def getBalance(inputs, reference_inputs, parameters):
    user = int(parameters[0])
    filepath = parameters[1]
    outputs = ()
    try:
        fp = open(filepath, "a+")
        for index in range(3):
            line = fp.next()
            balance[index] = int(line.strip())
    except StopIteration:
        fp.close()
    else:
        fp.close()
    outputs = outputs + ("The balance of user {0} is {1}.".format(user, balance[user]),)
    return {
        'outputs': outputs
    }

@contract.method('transfer')
def transfer(inputs, reference_inputs, parameters):
    user_a, user_b, value, filepath = int(parameters[0]), int(parameters[1]), int(parameters[2]), parameters[3]
    outputs = ()
    try:
        fp = open(filepath, "a+")
        for index in range(3):
            line = fp.next()
            balance[index] = int(line.strip())
    except StopIteration:
        fp.close()
    else:
        fp.close()
    try:
        wfp = open(filepath, "w")
    except IOError:
        raise IOError("Cannot open write file")
    else:
        if balance[user_a] < value:
            outputs = outputs + ("user {0} has not enough coins!\n".format(user_a),)
        else:
            balance[user_a] -= value
            balance[user_b] += value
            for i in range(3):
                wfp.write("{0}\n".format(balance[i]))
            outputs = outputs + ("ok\n",)
        wfp.close()
    return {
        'outputs': outputs
    }

if __name__ == '__main__':
    contract.run()
